Mathlib/Tactic/Widget/StringDiagram.lean

```diff
@@ -0,0 +1,363 @@
+/-
+Copyright (c) 2024 Yuma Mizuno. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yuma Mizuno
+-/
+import ProofWidgets.Component.PenroseDiagram
+import ProofWidgets.Presentation.Expr
+import Mathlib.Tactic.CategoryTheory.Monoidal
+
+/-!
+# String Diagram Widget
+
+This file provides meta infrastructure for displaying string diagrams for morphisms in monoidal
+categories in the infoview. To enable the string diagram widget, you need to import this file and
+inserting `with_panel_widgets [Mathlib.Tactic.Widget.StringDiagram]` at the beginning of the
+proof. Alternatively, you can also write
+```lean
+open Mathlib.Tactic.Widget
+show_panel_widgets [local StringDiagram]
+```
+to enable the string diagram widget in the current section.
+
+String diagrams are graphical representations of morphisms in monoidal categories, which are
+useful for rewriting computations. More precisely, objects in a monoidal category is represented
+by strings, and morphisms between two objects is represented by nodes connecting two strings
+associated with the objects. The tensor product `X ⊗ Y` corresponds to putting strings associated
+with `X` and `Y` horizontally (from left to right), and the composition of morphisms `f : X ⟶ Y`
+and `g : Y ⟶ Z` corresponds to connecting two nodes associated with `f` and `g` vertically (from
+top to bottom) by strings associated with `Y`.
+
+Currently, the string diagram widget provided in this file deals with equalities of morphisms
+in monoidal categories. It displays string diagrams corresponding to the morphisms for the
+left-hand and right-hand sides of the equality.
+
+Some examples can be found in `test/StringDiagram.lean`.
+
+When drawing string diagrams, it is common to ignore associators and unitors. We follow this
+convention. To do this, we need to extract non-structural morphisms that are not associators
+and unitors from lean expressions. This operation is performed using the `Tactic.Monoidal.eval`
+function.
+
+A monoidal category can be viewed as a bicategory with a single object. The program in this
+file can also be used to display the string diagram for general bicategories (see the wip
+PR #12107). With this in mind we will sometimes refer to objects and morphisms in monoidal
+categories as 1-morphisms and 2-morphisms respectively, borrowing the terminology of bicategories.
+Note that the relation between monoidal categories and bicategories is formalized in
+`Mathlib.CategoryTheory.Bicategory.SingleObj`, although the string diagram widget does not use
+it directly.
+
+-/
+
+namespace Mathlib.Tactic
+
+open Lean Meta Elab
+open CategoryTheory
+open Mathlib.Tactic.Coherence
+
+open Mathlib.Tactic.Monoidal
+
+namespace Widget.StringDiagram
+
+/-! ## Objects in string diagrams -/
+
+/-- Nodes for 2-morphisms in a string diagram. -/
+structure AtomNode : Type where
+  /-- The vertical position of the node in the string diagram. -/
+  vPos : ℕ
```

Comment:
I wonder if there is any value to using something unboxed like `UInt` in place of `Nat`.